File: PO PROJEKT 1/PO PROJEKT 1/PROJEKT PO PLIKI ŹRÓDŁOWE/Organism.cpp
#include "Organism.h"

void Organism::draw() {
	cout << this->sprite;
}

void Organism::spawn() {
	Coordinate coords;
    bool spawned = false;
    bool cycled = false;
    if (world->board.size() < world->getColumns() * world->getRows()) {
        coords.y = rand() % world->getRows();
        coords.x = rand() % world->getColumns();
        while (!spawned) { // jeœli wylosowane pole jest zajête wybierane jest nastêpne wolne pole
            if (!world->board.count(coords)) {
                world->board.insert({ coords, this });
                spawned = true;
            }
            else if (coords.x != world->getColumns()) {
                coords.x += 1;
            }
            else if (coords.y != world->getRows()) {
                coords.y += 1;
                coords.x = 0;
            }
            else {
                coords.x = 0;
                coords.y = 0;
            }
        }
    }
}

bool Organism::ranAway(Coordinate coord, Coordinate newCoord) {
	return false;
}

bool Organism::hasDeflectedAttack(Coordinate coord, Coordinate newCoord) {
    return false;
}

bool Organism::hasIncreasedStrength(Coordinate coord) {
    return false;
}

bool Organism::poisonous(Coordinate coord, Coordinate newCoord) {
    return false;
}


Coordinate Organism::checkSurroundings(Coordinate coord) { // w przypadku gdy nie ma wolnego pola, funkcja zwraca niezmieniony koordynat
	Coordinate newCoord = coord;
    bool nClear = true, eClear = true, wClear = true, sClear = true;
    while (nClear || eClear || wClear || sClear) {
        int dir = rand() % 4;
        if (dir == N && nClear == true) {
            newCoord.y -= 1;
            if (newCoord.y < 0) {
                newCoord.y += 1;
                nClear = false;
            }
            else if (!world->board.count(newCoord)) {
                return newCoord;
            }
            else {
                newCoord.y += 1;
                nClear = false;
                dir++;
            }
        }
        if (dir == E && eClear == true) {
            newCoord.x += 1;
            if (newCoord.x >= world->getColumns() || world->board.count(newCoord)) {
                newCoord.x -= 1;
                eClear = false;
            }
            else if (!world->board.count(newCoord)) {
                return newCoord;
            }
            else {
                newCoord.x -= 1;
                eClear = false;
                dir++;
            }

        }
        if (dir == S && sClear == true) {
            newCoord.y += 1;
            if (newCoord.y >= world->getRows() || world->board.count(newCoord)) {
                newCoord.y -= 1;
                sClear = false;
                dir++;
            }
            else if (!world->board.count(newCoord)) {
                return newCoord;
            }
        }
        if (dir == W && wClear == true) {
            newCoord.x -= 1;
            if (newCoord.x < 0 || world->board.count(newCoord)) {
                newCoord.x += 1;
                wClear = false;
            }
            else if (!world->board.count(newCoord)) {
                return newCoord;
            }
        }
    }
    return coord;
}

void Organism::endTurn() {
    age += 1;
}

Coordinate Organism::findCoord() {
    for (auto it = world->board.cbegin(), next_it = it; it != world->board.cend(); it = next_it) {
        if (this == it->second) {
            return it->first;
        }
        ++next_it;
    }
    Coordinate failedCoord;
    failedCoord.x = -1, failedCoord.y = -1;
    return failedCoord;
}

void Organism::multiply(Coordinate coord) {
    Coordinate childCoord = checkSurroundings(coord); // metoda checkSurroundings w przypadku braku wolnego pola w s¹siedztwie zwraca niezmienione pole
    if (childCoord == coord) {

    }
    else {
        Organism* child = createChild(childCoord);
    }
}

int Organism::getInitiative() {
    return this->initiative;
}

int Organism::getStrength() {
    return this->strength;
}

int Organism::getAge() {
    return this->age;
}

int Organism::getSuperSpeedTimer() {
    return NULL;
}

void Organism::setInitiative(int initiative) {
    this->initiative = initiative;
}

void Organism::increaseStrength(int strength) {
    this->strength += strength;
}

char Organism::getSprite() {
    return sprite;
}

void Organism::setAge(int age) {
    this->age = age;
}

File: Symulator Świata/organism.py
from coordinates import Coordinates
import directions
import random
from abc import abstractmethod


class Organism:

    def __init__(self, orgWorld, coord=None):
        self._strength = 0
        self._initiative = 0
        self._age = 0
        self._sprite = None
        self.world = orgWorld

    @abstractmethod
    def action(self, coord):
        pass

    def endTurn(self):
        self._age += 1

    @property
    def initiative(self):
        return self._initiative

    @property
    def strength(self):
        return self._strength

    @property
    def age(self):
        return self._age

    @initiative.setter
    def initiative(self, value):
        self._initiative = value

    @strength.setter
    def strength(self, value):
        self._strength = value

    @age.setter
    def age(self, value):
        self._age = value

    def ranAway(self, coord, newCoord):
        return False

    def hasDeflectedAttack(self, coord, newCoord):
        return False

    def hasIncreasedStrength(self, coord):
        return False

    def poisonous(self, coord, newCoord):
        return False

    def poisonResistant(self):
        return False

    def spawn(self):
        if len(self.world.board) <= self.world.getColumns() * self.world.getRows():
            while True:
                coord = Coordinates(random.randint(0, self.world.getRows()-1), random.randint(0, self.world.getColumns() - 1))
                if coord not in self.world.board:
                    self.world.board[coord] = self
                    break

    def checkSurroundings(self, coord):
        x = coord.x
        y = coord.y
        newCoord = Coordinates(x, y)
        nClear = True
        eClear = True
        wClear = True
        sClear = True
        dir = random.randint(0, 3)
        while nClear & eClear & wClear & sClear:
            if dir == directions.N:
                newCoord.y -= 1
                if newCoord.y < 0:
                    newCoord.y += 1
                    nClear = False
                elif newCoord not in self.world.board.keys():
                    return newCoord
                else:
                    newCoord.y += 1
                    nClear = False
                    dir += 1
            if dir == directions.E:
                newCoord.x += 1
                if newCoord.x >= self.world.getColumns():
                    newCoord.x -= 1
                    eClear = False
                elif newCoord not in self.world.board.keys():
                    return newCoord
                else:
                    newCoord.x -= 1
                    eClear = False
                    dir += 1
            if dir == directions.S:
                newCoord.y += 1
                if newCoord.y >= self.world.getRows():
                    newCoord.y -= 1
                    sClear = False
                elif newCoord not in self.world.board.keys():
                    return newCoord
                else:
                    newCoord.y -= 1
                    sClear = False
                    dir += 1
            if dir == directions.W:
                newCoord.x -= 1
                if newCoord.x < 0:
                    newCoord.x += 1
                    wClear = False
                elif newCoord not in self.world.board.keys():
                    return newCoord
                else:
                    newCoord.x += 1
                    wClear = False
                    dir = 0
        return coord

    def multiply(self, coord):
        childCoord = self.checkSurroundings(coord)
        if childCoord != coord:
            child = self.createChild(childCoord)

    @abstractmethod
    def createChild(self, coord):
        pass

    def findCoordinates(self):
        if self in self.world.board.values():
            for key in self.world.board:
                if key in self.world.board.keys():
                    if self is self.world.board[key]:
                        return key
